File: cpp_tetris/src/Tetris.cpp
#include "tetris/Tetris.h"
#include <vector>
#include <memory>
#include <functional>

namespace tetris {

Tetris::Tetris(int width, int height) 
    : gameRunning_(true), score_(0), gen_(rd_()), dis_(0, 3) {
    console_ = std::make_shared<Console>();
    board_ = std::make_unique<Board>(width, height);
    boardView_ = std::make_shared<BoardView>(console_);
}

void Tetris::spawnNewPiece() {
    // Array of piece creation functions
    std::vector<std::function<std::shared_ptr<Piece>()>> pieceCreators = {
        PieceFactory::createIPiece,
        PieceFactory::createOPiece,
        PieceFactory::createTPiece,
        PieceFactory::createLPiece
    };

    std::vector<std::function<std::unique_ptr<PieceView>()>> pieceViewCreators = {
        PieceFactory::createIPieceView,
        PieceFactory::createOPieceView,
        PieceFactory::createTPieceView,
        PieceFactory::createLPieceView
    };

    int randomIndex = dis_(gen_);
    board_->setCurrentPiece(pieceCreators[randomIndex]());
    currentPieceView_ = pieceViewCreators[randomIndex]();
}

void Tetris::processUserInput(const std::string& input) {
    if (!input.empty()) {
        char command = input[0];
        auto currentPiece = board_->getCurrentPiece();
        
        if (!currentPiece) return;

        switch (command) {
            case '4': // Move left
                if (board_->canMovePiece(*currentPiece, -1, 0)) {
                    currentPiece->moveLeft();
                }
                break;
            case '6': // Move right
                if (board_->canMovePiece(*currentPiece, 1, 0)) {
                    currentPiece->moveRight();
                }
                break;
            case '7': // Rotate counter-clockwise
                board_->canRotatePiece(*currentPiece, false);
                break;
            case '9': // Rotate clockwise
                board_->canRotatePiece(*currentPiece, true);
                break;
        }
    }
}

void Tetris::gameLoop() {
    spawnNewPiece();
    
    while (gameRunning_) {
        boardView_->display(*board_, currentPieceView_.get());
        
        std::string prompt = "Score: " + std::to_string(score_) + 
                           "\n\nComando (4=izq, 6=der, 7=rotar↺, 9=rotar↻): ";
        std::string input = console_->readString(prompt);
        
        processUserInput(input);

        auto currentPiece = board_->getCurrentPiece();
        if (!currentPiece) {
            continue;
        }

        // Try to move piece down
        if (board_->canMovePiece(*currentPiece, 0, 1)) {
            currentPiece->moveDown();
        } else {
            // Piece can't move down, place it
            board_->placePiece(*currentPiece, currentPieceView_->getSymbol());

            // Clear complete lines
            int linesCleared = board_->clearCompleteLines();
            score_ += linesCleared * 100;

            if (linesCleared > 0) {
                console_->writeln("¡" + std::to_string(linesCleared) + " línea(s) eliminada(s)!");
                console_->readString("Presiona Enter para continuar...");
            }

            // Spawn new piece
            spawnNewPiece();

            // Check game over
            if (!board_->canMovePiece(*board_->getCurrentPiece(), 0, 0)) {
                gameRunning_ = false;
                console_->writeln("¡GAME OVER!");
                console_->writeln("Puntuación final: " + std::to_string(score_));
            }
        }
    }
}

} // namespace tetris

File: python_src/tetris.py
"""Tetris game main logic."""

import random
from typing import Optional
from .board import Board
from .board_view import BoardView
from .console import Console
from .piece import Piece
from .piece_view import PieceView
from .piece_factory import PieceFactory


class Tetris:
    """Main Tetris game controller."""
    
    def __init__(self, width: int, height: int):
        """Initialize the Tetris game.
        
        Args:
            width: Board width
            height: Board height
        """
        self.console = Console()
        self.board = Board(width, height)
        self.board_view = BoardView(self.console)
        self.game_running = True
        self.score = 0
        self.current_piece_view: Optional[PieceView] = None
    
    def spawn_new_piece(self) -> None:
        """Spawn a new random piece at the top of the board."""
        # Available pieces and their views
        pieces = [
            PieceFactory.create_i_piece(),
            PieceFactory.create_o_piece(),
            PieceFactory.create_t_piece(),
            PieceFactory.create_l_piece()
        ]
        
        piece_views = [
            PieceFactory.create_i_piece_view(),
            PieceFactory.create_o_piece_view(),
            PieceFactory.create_t_piece_view(),
            PieceFactory.create_l_piece_view()
        ]
        
        # Randomly select a piece
        random_index = random.randint(0, len(pieces) - 1)
        self.board.set_current_piece(pieces[random_index])
        self.current_piece_view = piece_views[random_index]
    
    def process_user_input(self, user_input: str) -> None:
        """Process user input commands.
        
        Args:
            user_input: User input string
        """
        if not user_input:
            return
        
        command = user_input[0]
        current_piece = self.board.get_current_piece()
        
        if not current_piece:
            return
        
        if command == '4':  # Move left
            if self.board.can_move_piece(current_piece, -1, 0):
                current_piece.move_left()
        elif command == '6':  # Move right
            if self.board.can_move_piece(current_piece, 1, 0):
                current_piece.move_right()
        elif command == '7':  # Rotate counter-clockwise
            self.board.can_rotate_piece(current_piece, False)
        elif command == '9':  # Rotate clockwise
            self.board.can_rotate_piece(current_piece, True)
    
    def game_loop(self) -> None:
        """Main game loop."""
        self.spawn_new_piece()
        
        while self.game_running:
            # Display the board
            self.board_view.display(self.board, self.current_piece_view)
            
            # Get user input
            prompt = (f"Score: {self.score}\\n\\n"
                     "Comando (4=izq, 6=der, 7=rotar↺, 9=rotar↻): ")
            user_input = self.console.read_string(prompt)
            
            # Process user input
            self.process_user_input(user_input)
            
            current_piece = self.board.get_current_piece()
            if not current_piece:
                continue
            
            # Try to move piece down
            if self.board.can_move_piece(current_piece, 0, 1):
                current_piece.move_down()
            else:
                # Piece can't move down, place it permanently
                self.board.place_piece(current_piece, 
                                     self.current_piece_view.get_symbol())
                
                # Clear complete lines
                lines_cleared = self.board.clear_complete_lines()
                self.score += lines_cleared * 100
                
                # Show line clear message
                if lines_cleared > 0:
                    self.console.writeln(f"¡{lines_cleared} línea(s) eliminada(s)!")
                    self.console.read_string("Presiona Enter para continuar...")
                
                # Spawn new piece
                self.spawn_new_piece()
                
                # Check game over
                new_piece = self.board.get_current_piece()
                if new_piece and not self.board.can_move_piece(new_piece, 0, 0):
                    self.game_running = False
                    self.console.writeln("¡GAME OVER!")
                    self.console.writeln(f"Puntuación final: {self.score}")